File: android/app/src/main/java/com/poshanforlife/android/feature/practitioner/upload/ReportFieldGroups.kt
package com.poshanforlife.android.feature.practitioner.upload

import com.poshanforlife.android.core.network.InBodyDataDto

data class FieldSpec(val key: String, val label: String, val unit: String, val isInt: Boolean = false)

data class FieldGroupSpec(val title: String, val fields: List<FieldSpec>)

/**
 * Backend's InBodyData is a flat 20-field scaffold with no server-side
 * grouping (see InBodyData.java) — these groups follow the section names
 * from the original Expo/web spec ("Body Composition, Muscle-Fat, Obesity,
 * Score, Segmental, Impedance"), best-fit-mapped onto the 20 fields that
 * actually exist. "Segmental" (per-limb lean mass) is omitted entirely —
 * no backend field models it; add a group here if a future backend prompt
 * extends InBodyData with real segmental data. "Impedance" holds
 * intracellular/extracellular water as the closest available proxy, since
 * there's no raw per-segment impedance (ohm) reading in the schema either.
 */
val REPORT_FIELD_GROUPS: List<FieldGroupSpec> = listOf(
    FieldGroupSpec(
        title = "Body Composition",
        fields = listOf(
            FieldSpec("weightKg", "Weight", "kg"),
            FieldSpec("bodyWaterL", "Total body water", "L"),
            FieldSpec("proteinKg", "Protein", "kg"),
            FieldSpec("mineralKg", "Mineral", "kg"),
            FieldSpec("bodyFatMassKg", "Body fat mass", "kg"),
            FieldSpec("fatFreeMassKg", "Fat-free mass", "kg"),
        ),
    ),
    FieldGroupSpec(
        title = "Muscle-Fat",
        fields = listOf(
            FieldSpec("skeletalMuscleMassKg", "Skeletal muscle mass", "kg"),
            FieldSpec("bodyFatPercent", "Body fat %", "%"),
        ),
    ),
    FieldGroupSpec(
        title = "Obesity",
        fields = listOf(
            FieldSpec("bmi", "BMI", ""),
            FieldSpec("waistHipRatio", "Waist-hip ratio", ""),
            FieldSpec("visceralFatLevel", "Visceral fat level", ""),
            FieldSpec("obesityDegreePercent", "Obesity degree", "%"),
            FieldSpec("targetWeightKg", "Target weight", "kg"),
            FieldSpec("weightControlKg", "Weight control", "kg"),
            FieldSpec("fatControlKg", "Fat control", "kg"),
            FieldSpec("muscleControlKg", "Muscle control", "kg"),
        ),
    ),
    FieldGroupSpec(
        title = "Score",
        fields = listOf(
            FieldSpec("inbodyScore", "InBody score", "", isInt = true),
            FieldSpec("basalMetabolicRate", "Basal metabolic rate", "kcal"),
        ),
    ),
    FieldGroupSpec(
        title = "Impedance",
        fields = listOf(
            FieldSpec("intracellularWaterL", "Intracellular water", "L"),
            FieldSpec("extracellularWaterL", "Extracellular water", "L"),
        ),
    ),
)

/** Total field count backing the confidence-tier thresholds in ReportUploadViewModel. */
val REPORT_FIELD_COUNT: Int = REPORT_FIELD_GROUPS.sumOf { it.fields.size }

fun InBodyDataDto.valueFor(key: String): Double? = when (key) {
    "weightKg" -> weightKg
    "bodyFatPercent" -> bodyFatPercent
    "skeletalMuscleMassKg" -> skeletalMuscleMassKg
    "bmi" -> bmi
    "visceralFatLevel" -> visceralFatLevel
    "bodyWaterL" -> bodyWaterL
    "proteinKg" -> proteinKg
    "mineralKg" -> mineralKg
    "basalMetabolicRate" -> basalMetabolicRate
    "bodyFatMassKg" -> bodyFatMassKg
    "fatFreeMassKg" -> fatFreeMassKg
    "waistHipRatio" -> waistHipRatio
    "targetWeightKg" -> targetWeightKg
    "weightControlKg" -> weightControlKg
    "fatControlKg" -> fatControlKg
    "muscleControlKg" -> muscleControlKg
    "obesityDegreePercent" -> obesityDegreePercent
    "intracellularWaterL" -> intracellularWaterL
    "extracellularWaterL" -> extracellularWaterL
    "inbodyScore" -> inbodyScore?.toDouble()
    else -> null
}

fun InBodyDataDto.withValue(key: String, value: Double?): InBodyDataDto = when (key) {
    "weightKg" -> copy(weightKg = value)
    "bodyFatPercent" -> copy(bodyFatPercent = value)
    "skeletalMuscleMassKg" -> copy(skeletalMuscleMassKg = value)
    "bmi" -> copy(bmi = value)
    "visceralFatLevel" -> copy(visceralFatLevel = value)
    "bodyWaterL" -> copy(bodyWaterL = value)
    "proteinKg" -> copy(proteinKg = value)
    "mineralKg" -> copy(mineralKg = value)
    "basalMetabolicRate" -> copy(basalMetabolicRate = value)
    "bodyFatMassKg" -> copy(bodyFatMassKg = value)
    "fatFreeMassKg" -> copy(fatFreeMassKg = value)
    "waistHipRatio" -> copy(waistHipRatio = value)
    "targetWeightKg" -> copy(targetWeightKg = value)
    "weightControlKg" -> copy(weightControlKg = value)
    "fatControlKg" -> copy(fatControlKg = value)
    "muscleControlKg" -> copy(muscleControlKg = value)
    "obesityDegreePercent" -> copy(obesityDegreePercent = value)
    "intracellularWaterL" -> copy(intracellularWaterL = value)
    "extracellularWaterL" -> copy(extracellularWaterL = value)
    "inbodyScore" -> copy(inbodyScore = value?.toInt())
    else -> this
}

fun InBodyDataDto.nonNullFieldCount(): Int =
    REPORT_FIELD_GROUPS.sumOf { group -> group.fields.count { valueFor(it.key) != null } }

File: ios/PoshanForLife/Features/Practitioner/Upload/ReportFieldGroups.swift
import Foundation

struct FieldSpec: Identifiable {
    let key: String
    let label: String
    let unit: String
    var isInt: Bool = false
    var id: String { key }
}

struct FieldGroupSpec: Identifiable {
    let title: String
    let fields: [FieldSpec]
    var id: String { title }
}

/// Backend's `InBodyData` is a flat 20-field scaffold with no server-side
/// grouping — these groups follow the section names from the original
/// Expo/web spec ("Body Composition, Muscle-Fat, Obesity, Score, Segmental,
/// Impedance"), best-fit-mapped onto the 20 fields that actually exist.
/// "Segmental" (per-limb lean mass) is omitted entirely — no backend field
/// models it. "Impedance" holds intracellular/extracellular water as the
/// closest available proxy. Mirrors Android's `REPORT_FIELD_GROUPS` exactly,
/// field-for-field, so the two clients group the same 20 values the same way.
enum ReportFieldGroups {
    static let all: [FieldGroupSpec] = [
        FieldGroupSpec(title: "Body Composition", fields: [
            FieldSpec(key: "weightKg", label: "Weight", unit: "kg"),
            FieldSpec(key: "bodyWaterL", label: "Total body water", unit: "L"),
            FieldSpec(key: "proteinKg", label: "Protein", unit: "kg"),
            FieldSpec(key: "mineralKg", label: "Mineral", unit: "kg"),
            FieldSpec(key: "bodyFatMassKg", label: "Body fat mass", unit: "kg"),
            FieldSpec(key: "fatFreeMassKg", label: "Fat-free mass", unit: "kg"),
        ]),
        FieldGroupSpec(title: "Muscle-Fat", fields: [
            FieldSpec(key: "skeletalMuscleMassKg", label: "Skeletal muscle mass", unit: "kg"),
            FieldSpec(key: "bodyFatPercent", label: "Body fat %", unit: "%"),
        ]),
        FieldGroupSpec(title: "Obesity", fields: [
            FieldSpec(key: "bmi", label: "BMI", unit: ""),
            FieldSpec(key: "waistHipRatio", label: "Waist-hip ratio", unit: ""),
            FieldSpec(key: "visceralFatLevel", label: "Visceral fat level", unit: ""),
            FieldSpec(key: "obesityDegreePercent", label: "Obesity degree", unit: "%"),
            FieldSpec(key: "targetWeightKg", label: "Target weight", unit: "kg"),
            FieldSpec(key: "weightControlKg", label: "Weight control", unit: "kg"),
            FieldSpec(key: "fatControlKg", label: "Fat control", unit: "kg"),
            FieldSpec(key: "muscleControlKg", label: "Muscle control", unit: "kg"),
        ]),
        FieldGroupSpec(title: "Score", fields: [
            FieldSpec(key: "inbodyScore", label: "InBody score", unit: "", isInt: true),
            FieldSpec(key: "basalMetabolicRate", label: "Basal metabolic rate", unit: "kcal"),
        ]),
        FieldGroupSpec(title: "Impedance", fields: [
            FieldSpec(key: "intracellularWaterL", label: "Intracellular water", unit: "L"),
            FieldSpec(key: "extracellularWaterL", label: "Extracellular water", unit: "L"),
        ]),
    ]
}

extension InBodyData {
    func value(for key: String) -> Double? {
        switch key {
        case "weightKg": return weightKg
        case "bodyFatPercent": return bodyFatPercent
        case "skeletalMuscleMassKg": return skeletalMuscleMassKg
        case "bmi": return bmi
        case "visceralFatLevel": return visceralFatLevel
        case "bodyWaterL": return bodyWaterL
        case "proteinKg": return proteinKg
        case "mineralKg": return mineralKg
        case "basalMetabolicRate": return basalMetabolicRate
        case "bodyFatMassKg": return bodyFatMassKg
        case "fatFreeMassKg": return fatFreeMassKg
        case "waistHipRatio": return waistHipRatio
        case "targetWeightKg": return targetWeightKg
        case "weightControlKg": return weightControlKg
        case "fatControlKg": return fatControlKg
        case "muscleControlKg": return muscleControlKg
        case "obesityDegreePercent": return obesityDegreePercent
        case "intracellularWaterL": return intracellularWaterL
        case "extracellularWaterL": return extracellularWaterL
        case "inbodyScore": return inbodyScore.map(Double.init)
        default: return nil
        }
    }

    mutating func setValue(_ value: Double?, for key: String) {
        switch key {
        case "weightKg": weightKg = value
        case "bodyFatPercent": bodyFatPercent = value
        case "skeletalMuscleMassKg": skeletalMuscleMassKg = value
        case "bmi": bmi = value
        case "visceralFatLevel": visceralFatLevel = value
        case "bodyWaterL": bodyWaterL = value
        case "proteinKg": proteinKg = value
        case "mineralKg": mineralKg = value
        case "basalMetabolicRate": basalMetabolicRate = value
        case "bodyFatMassKg": bodyFatMassKg = value
        case "fatFreeMassKg": fatFreeMassKg = value
        case "waistHipRatio": waistHipRatio = value
        case "targetWeightKg": targetWeightKg = value
        case "weightControlKg": weightControlKg = value
        case "fatControlKg": fatControlKg = value
        case "muscleControlKg": muscleControlKg = value
        case "obesityDegreePercent": obesityDegreePercent = value
        case "intracellularWaterL": intracellularWaterL = value
        case "extracellularWaterL": extracellularWaterL = value
        case "inbodyScore": inbodyScore = value.map { Int($0) }
        default: break
        }
    }
}
